### ISSUE TITLE
Accuracy problem of torch.batch_norm_gather_stats_with_counts when running_mean is half tensor

### ISSUE THREAD

**Issue by cdzhan** [pytorch/pytorch#70901]
### 🐛 Describe the bug

The value of output tensor "mea" should not be zero.

```
import torch
i = torch.rand([5,4,10,10], dtype=torch.float).cuda()
c = torch.rand((150,), dtype=torch.half).fill_(500).cuda()
m = torch.rand([150, 4], dtype=torch.float).cuda()
rm = torch.rand([4], dtype=torch.half).cuda()
v = torch.rand([150, 4], dtype=torch.float).cuda()
v += 0.0001
rv = torch.rand([4], dtype=torch.half).cuda()
rv += 0.0001
mea, inv = torch.batch_norm_gather_stats_with_counts(i, m, v, rm, rv, 1.2, 0.001, c)
print(mea)
```
output:

```
tensor([0., 0., 0., 0.], device='cuda:0')
```

### Versions

PyTorch version: 1.10.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
GCC version: (GCC) 5.4.0
Clang version: Could not collect
CMake version: version 3.15.5
Libc version: glibc-2.3.4

Python version: 3.6.8 (default, Nov 16 2020, 16:55:22)  [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] (64-bit runtime)
Python platform: Linux-3.10.0-1160.15.2.el7.x86_64-x86_64-with-centos-7.9.2009-Core
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB

Nvidia driver version: 440.64.00
cuDNN version: Probably one of the following:
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] torch==1.10.1
[conda] Could not collect

**Comment by albanD** [comment]
Isn't this expected behavior due to lower precision of fp16: https://pytorch.org/docs/master/notes/numerical_accuracy.html ?

**Comment by cdzhan** [comment]
I think the root cause is the lower precision of fp16, but I do not think it is the expected behavior. Why not output more accurate result by just simply using higher precision data type (e.g., float) in the middle step of caculation?

**Comment by cdzhan** [comment]
Another risk is that the dtype of `counts ` is follow  `mean`, but they are in different dtype when using GenericPackedTensorAccessor in the kernel `batch_norm_gather_stats_cuda_template`. It will throw error when `running_mean` is in different dtype with `mean`.

**Comment by cdzhan** [comment]
@ngimel @jjsjann123 @albanD Would you like to take a look at this issue?
